[view-transitions] Implement "update pseudo-element styles" algorithm.

mattwoodrow opened the pull request:
#### 810e27d205f59629bff16d325ae44e5c4d2ea5e4
<pre>
[view-transitions] Implement &quot;update pseudo-element styles&quot; algorithm.
<a href="https://bugs.webkit.org/show_bug.cgi?id=265224">https://bugs.webkit.org/show_bug.cgi?id=265224</a>
<a href="https://rdar.apple.com/118703293">rdar://118703293</a>&gt;

Reviewed by Tim Nguyen.

Implements the basic outline of capturing the computed style from both
the old and new elements, and using these to generate a style rule
for the selector &apos;:root::view-transition-group(transitionName)&apos;.

Adds a mostly empty RenderViewTransitionOld class that we will use
for the replaced snapshot content, but just acts as a placeholder
for the intrinsic size right now.

With these changes, the size of the transitioned element now gets
applied correctly to the :view-transition-new/old renderers, and
styles targetting them (like background color) correctly get
applied and rendered.

The next step is to implement the &apos;capture the image&apos; algorithm
for new and old (likely splitting RenderViewTransformNew as a separate
class), so that we also include the replaced content.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::ViewTransition::captureOldState):
(WebCore::ViewTransition::handleTransitionFrame):
(WebCore::ViewTransition::copyElementBaseProperties):
(WebCore::ViewTransition::updatePseudoElementStyles):
* Source/WebCore/dom/ViewTransition.h:
(WebCore::OrderedNamedElementsMap::map const):
(WebCore::OrderedNamedElementsMap::find const):
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isRenderViewTransitionOld const):
* Source/WebCore/rendering/RenderViewTransitionOld.cpp: Added.
(WebCore::RenderViewTransitionOld::RenderViewTransitionOld):
(WebCore::RenderViewTransitionOld::setImage):
* Source/WebCore/rendering/RenderViewTransitionOld.h: Added.
* Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
(WebCore::RenderTreeUpdater::ViewTransition::buildPseudoElementGroup):
(WebCore::RenderTreeUpdater::ViewTransition::updatePseudoElementGroup):
* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::ElementRuleCollector::ElementRuleCollector):
(WebCore::Style::ElementRuleCollector::matchUARules):
* Source/WebCore/style/ElementRuleCollector.h:
* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::applyMatchedProperties):
(WebCore::Style::Resolver::setViewTransitionGroupStyles):
* Source/WebCore/style/StyleResolver.h:
* Source/WebCore/style/StyleScope.cpp:
(WebCore::Style::Scope::createDocumentResolver):
* Source/WebCore/style/StyleScope.h:
* Source/WebCore/style/StyleScopeRuleSets.cpp:
(WebCore::Style::ScopeRuleSets::dynamicViewTransitionsStyle const):
* Source/WebCore/style/StyleScopeRuleSets.h:
(WebCore::Style::ScopeRuleSets::setDynamicViewTransitionsStyle):

Canonical link: <a href="https://commits.webkit.org/274817@main">https://commits.webkit.org/274817@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2ae504dead819a7c847828a3f19f6011ccc8c473

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40097 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19109 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42474 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42642 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/36186 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42404 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22022 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16438 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33363 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40671 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16090 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34635 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13919 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/13951 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35584 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43920 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36378 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35919 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39679 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/14912 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12237 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/37951 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16531 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9000 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16580 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/16175 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -349,7 +359,7 @@ void ViewTransition::handleTransitionFrame()
 
     // FIXME: If transition’s initial snapshot containing block size is not equal to the snapshot containing block size, then skip the view transition for transition, and return.
-    // FIXME: Update pseudo-element styles for transition.
+    updatePseudoElementStyles();
```
nit: the spec says to call this in  handleTransitionFrame() too

nt1m commented on Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
```diff
@@ -143,20 +156,33 @@ void RenderTreeUpdater::ViewTransition::buildPseudoElementGroup(const AtomString
+            rendererViewTransitionOld = WebCore::createRenderer<RenderViewTransitionOld>(RenderObject::Type::ViewTransitionOld, documentElementRenderer.document(), WTFMove(newStyle));
+
+            const CapturedElement* capturedElement = documentElementRenderer.document().activeViewTransition()->namedElements().find(name);
+            if (capturedElement)
```
```suggestion
            if (auto* capturedElement = documentElementRenderer.document().activeViewTransition()->namedElements().find(name);)
```

mdubet commented on Source/WebCore/dom/ViewTransition.h:
```diff
@@ -104,6 +107,13 @@ struct OrderedNamedElementsMap {
         return nullptr;
     }
 
+    const CapturedElement* find(const AtomString& key) const
```
We should return a `std::optional<CapturedElement>` or `const std::optional<CapturedElement>&` but not a pointer.

nt1m replied:
Is it how find methods are typically defined in WTF?

mdubet replied:
WTF uses iterators, so neither of this.

mdubet replied:
But well a pointer could be considered a kind of iterator. I'm fine with this :)

mdubet commented on Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
```diff
@@ -115,12 +116,24 @@ void RenderTreeUpdater::ViewTransition::buildPseudoElementGroup(const AtomString
+            RenderPtr<RenderViewTransitionOld> rendererViewTransitionOld;
+            rendererViewTransitionOld = WebCore::createRenderer<RenderViewTransitionOld>(RenderObject::Type::ViewTransitionOld, document, WTFMove(newStyle));
+
+            const CapturedElement* capturedElement = document->activeViewTransition()->namedElements().find(name);
```
`const auto& capturedElement = ...` (added bonus : it works with `optional`)

mdubet commented on Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
```diff
@@ -115,12 +116,24 @@ void RenderTreeUpdater::ViewTransition::buildPseudoElementGroup(const AtomString
+        RenderPtr<RenderBox> renderer;
+        if (pseudoId == PseudoId::ViewTransitionOld || pseudoId == PseudoId::ViewTransitionNew) {
+            RenderPtr<RenderViewTransitionOld> rendererViewTransitionOld;
+            rendererViewTransitionOld = WebCore::createRenderer<RenderViewTransitionOld>(RenderObject::Type::ViewTransitionOld, document, WTFMove(newStyle));
```
Merge with the previous line?

mdubet commented on Source/WebCore/style/StyleResolver.cpp:
```diff
@@ -702,5 +703,28 @@ std::optional<DynamicMediaQueryEvaluationChanges> Resolver::evaluateDynamicMedia
+    RuleSet* viewTransitionsStyle = m_ruleSets.dynamicViewTransitionsStyle();
+    RuleSetBuilder builder(*viewTransitionsStyle, mediaQueryEvaluator(), this);
+
+    if (!m_document)
```
Should we do this early return before to avoid creating variable for no reason?

nt1m commented on Source/WebCore/rendering/RenderViewTransition.h:
Can we name this `RenderViewTransitionCapture` to make it clear this is not the full view transition?

nt1m commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -396,4 +406,42 @@ void ViewTransition::clearViewTransition()
+// https://drafts.csswg.org/css-view-transitions-1/#update-pseudo-element-styles
+void ViewTransition::updatePseudoElementStyles()
+{
+    Style::Resolver& resolver = protectedDocument()->styleScope().resolver();
```
```suggestion
    auto& resolver = protectedDocument()->styleScope().resolver();
```

nt1m commented on Source/WebCore/style/StyleResolver.cpp:
```diff
@@ -707,5 +708,27 @@ std::optional<DynamicMediaQueryEvaluationChanges> Resolver::evaluateDynamicMedia
+    if (!m_document)
+        return;
+
+    RuleSet* viewTransitionsStyle = m_ruleSets.dynamicViewTransitionsStyle();
```
```suggestion
    auto* viewTransitionsStyle = m_ruleSets.dynamicViewTransitionsStyle();
```

nt1m commented on Source/WebCore/style/StyleResolver.cpp:
```diff
@@ -707,5 +708,27 @@ std::optional<DynamicMediaQueryEvaluationChanges> Resolver::evaluateDynamicMedia
+    RuleSetBuilder builder(*viewTransitionsStyle, mediaQueryEvaluator(), this);
+    CSSSelectorParserContext context(*m_document.get());
+
+    Vector<std::unique_ptr<MutableCSSSelector>> selectorList;
```
```suggestion
    MutableCSSSelectorList selectorList;
```

nt1m reviewed (commented):
This also needs to clear the dynamic styles when clearing the view transition.

nt1m commented on Source/WebCore/dom/ViewTransition.h:
```diff
@@ -52,12 +53,9 @@ struct CapturedElement {
-    // old mix-blend-mode
     WeakPtr<Element, WeakPtrImplWithEventTargetData> newElement;
 
     // FIXME: Also handle these:
```
The `group styles rule` can be removed from the FIXME.

mattwoodrow committed:
[view-transitions] Implement "update pseudo-element styles" algorithm.
https://bugs.webkit.org/show_bug.cgi?id=265224
rdar://118703293>

Reviewed by Tim Nguyen.

Implements the basic outline of capturing the computed style from both
the old and new elements, and using these to generate a style rule
for the selector ':root::view-transition-group(transitionName)'.

Adds a mostly empty RenderViewTransitionOld class that we will use
for the replaced snapshot content, but just acts as a placeholder
for the intrinsic size right now.

With these changes, the size of the transitioned element now gets
applied correctly to the :view-transition-new/old renderers, and
styles targetting them (like background color) correctly get
applied and rendered.

The next step is to implement the 'capture the image' algorithm
for new and old (likely splitting RenderViewTransformNew as a separate
class), so that we also include the replaced content.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::ViewTransition::captureOldState):
(WebCore::ViewTransition::handleTransitionFrame):
(WebCore::ViewTransition::copyElementBaseProperties):
(WebCore::ViewTransition::updatePseudoElementStyles):
* Source/WebCore/dom/ViewTransition.h:
(WebCore::OrderedNamedElementsMap::map const):
(WebCore::OrderedNamedElementsMap::find const):
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isRenderViewTransitionOld const):
* Source/WebCore/rendering/RenderViewTransitionOld.cpp: Added.
(WebCore::RenderViewTransitionOld::RenderViewTransitionOld):
(WebCore::RenderViewTransitionOld::setImage):
* Source/WebCore/rendering/RenderViewTransitionOld.h: Added.
* Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
(WebCore::RenderTreeUpdater::ViewTransition::buildPseudoElementGroup):
(WebCore::RenderTreeUpdater::ViewTransition::updatePseudoElementGroup):
* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::ElementRuleCollector::ElementRuleCollector):
(WebCore::Style::ElementRuleCollector::matchUARules):
* Source/WebCore/style/ElementRuleCollector.h:
* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::applyMatchedProperties):
(WebCore::Style::Resolver::setViewTransitionGroupStyles):
* Source/WebCore/style/StyleResolver.h:
* Source/WebCore/style/StyleScope.cpp:
(WebCore::Style::Scope::createDocumentResolver):
* Source/WebCore/style/StyleScope.h:
* Source/WebCore/style/StyleScopeRuleSets.cpp:
(WebCore::Style::ScopeRuleSets::dynamicViewTransitionsStyle const):
* Source/WebCore/style/StyleScopeRuleSets.h:
(WebCore::Style::ScopeRuleSets::setDynamicViewTransitionsStyle):

Canonical link: https://commits.webkit.org/274817@main
